src/main/scala/com/cognite/sdk/scala/v1/resources/dataModels.scala

```diff
@@ -60,28 +56,132 @@ class DataModels[F[_]](val requestSession: RequestSession[F])
+    Decoder
+      .decodeIterable[String, List]
+      .map {
+        case modelExternalId :: Nil =>
+          DataModelIdentifier(None, modelExternalId)
+        case spaceExternalId :: modelExternalId :: Nil =>
+          DataModelIdentifier(Some(spaceExternalId), modelExternalId)
+        case list =>
+          throw new SdkException(
+            s"Unable to decode DataModelIdentifier, expected array length 1 or 2, actual length ${list.length.toString}."
+          )
+      }
+  implicit val bTreeIndexDecoder: Decoder[BTreeIndex] =
+    deriveDecoder[BTreeIndex]
+  implicit val dataModelPropertyTypeDecoder: Decoder[PropertyType] =
+    Decoder.decodeString.map(
+      PropertyType
+        .fromCode(_)
+        .getOrElse(
+          throw new IllegalArgumentException("Invalid type specified")
+        )
+    )
+  implicit val dataModelPropertyIndexesDecoder: Decoder[DataModelIndexes] =
+    deriveDecoder[DataModelIndexes]
   implicit val dataModelPropertyDecoder: Decoder[DataModelProperty] =
     deriveDecoder[DataModelProperty]
-  implicit val dataModelDecoder: Decoder[DataModel] = deriveDecoder[DataModel]
-  implicit val dataModelItemsDecoder: Decoder[Items[DataModel]] = deriveDecoder[Items[DataModel]]
-
-  implicit val dataModelGetByCogniteIdIdsEncoder
-      : Encoder[DataModelGetByExternalIdsInput[CogniteId]] =
-    deriveEncoder[DataModelGetByExternalIdsInput[CogniteId]]
+  implicit val uniquenessConstraintDecoder: Decoder[UniquenessConstraint] =
+    new Decoder[UniquenessConstraint] {
+      final def apply(c: HCursor): Decoder.Result[UniquenessConstraint] =
+        for {
+          properties <- c.downField("uniqueProperties").as[Seq[Map[String, String]]]
+        } yield new UniquenessConstraint(properties.map(_("property")))
+    }
+  implicit val dataModelConstraintsDecoder: Decoder[DataModelConstraints] =
+    deriveDecoder[DataModelConstraints]
+  implicit val dataModelInstanceTypeDecoder: Decoder[DataModelInstanceType] =
+    deriveDecoder[DataModelInstanceType]
 
-  implicit val dataModelGetByExternalIdsEncoder
-      : Encoder[DataModelGetByExternalIdsInput[CogniteExternalId]] =
-    deriveEncoder[DataModelGetByExternalIdsInput[CogniteExternalId]]
+  // Gets the instanceType from allowNode and allowEdge, encodes it in a way the derived
+  //   decoder will understand, and adds it to the json before parsing
+  implicit val dataModelDecoder: Decoder[DataModel] =
+    deriveDecoder[DataModel].prepare { init =>
+      val allowNode = init.downField("allowNode").as[Boolean]
+      val allowEdge = init.downField("allowEdge").as[Boolean]
+      val instanceType: DataModelInstanceType = (allowNode, allowEdge) match {
+        case (Right(true), Right(false) | Left(_)) => DataModelInstanceType.Node
+        case (Right(false) | Left(_), Right(true)) => DataModelInstanceType.Edge
+        case _ => throw new Exception("Exactly one of allowNode and allowEdge must be true")
```

Comment:
Very nice solution! Shouldn't be necessary of course, and I think recent discussions might lead to a change in the API here.. but anyways, it would be nice to have a test or two for this particular decoding. Anyways, not clear to me if our current tests are testing this? I guess they have to test at least one of the paths to even read a Model, but let's make sure we at least test for (true false) and (false true)